### ISSUE TITLE
Why setting lr=0 for first 100 steps

### ISSUE THREAD

**Issue by BaohaoLiao** [ictnlp/PTE-NMT#1]
Hi, 

this is a great job. 

I have a question when I read your code. It seems you set lr=0 for the first 100 steps for both knowledge distillation and finetuning, may I ask why?

https://github.com/ictnlp/PTE-NMT/blob/6942753c5423728b9c4dc9600d052921ec0b4850/fairseq/tasks/fairseq_task.py#L287

**Comment by gushu333** [comment]
Hi，

This is because I found that the first updates (less than 100 steps) always change the values of the important parameters that should be frozen, which may be a bug of fairseq. So I set the learning rate to 0 at first to avoid this. 

This has nothing to do with the method itself and has no impact on the method's performance.

Shuhao Gu

**Comment by BaohaoLiao** [comment]
I think it do have impact on the unimportant parameter updates. 

You set the lr=0 for the first 100 steps. Even though all parameters are not updated, the first and second moments are updated. This could bring benefit to your method according to the paper Radam (first two paragraphs in section 3 at https://arxiv.org/pdf/1908.03265.pdf), making the training more stable without warmup (I notice you also don't use warmup for knowledge distillation or finetuning).

**Comment by BaohaoLiao** [comment]
Besides, I think the important parameters is updated because you do the back-propagation first and then set the gradient to 0. In this case, the first step still use the initial lr to update all parameters. Why don't you just set the values of important parameters to their original values after back-propagation without caring for their gradients? 

https://github.com/ictnlp/PTE-NMT/blob/6942753c5423728b9c4dc9600d052921ec0b4850/fairseq/tasks/fairseq_task.py#L284

**Comment by gushu333** [comment]
> I think it do have impact on the unimportant parameter updates.
> 
> You set the lr=0 for the first 100 steps. Even though all parameters are not updated, the first and second moments are updated. This could bring benefit to your method according to the paper Radam (first two paragraphs in section 3 at https://arxiv.org/pdf/1908.03265.pdf), making the training more stable without warmup (I notice you also don't use warmup for knowledge distillation or finetuning).

Sorry for the late reply, I forgot this at that time. 

The updating of the moments benefits the training process is very interesting and makes sense. I ignored this and thanking you for pointing this out!

**Comment by gushu333** [comment]
> Besides, I think the important parameters is updated because you do the back-propagation first and then set the gradient to 0. In this case, the first step still use the initial lr to update all parameters. Why don't you just set the values of important parameters to their original values after back-propagation without caring for their gradients?
> 
> https://github.com/ictnlp/PTE-NMT/blob/6942753c5423728b9c4dc9600d052921ec0b4850/fairseq/tasks/fairseq_task.py#L284

" the first step still use the initial lr to update all parameters" 
I am not so sure about this. As far as I am concerned, the update of the parameters should happen after the backward process and the resetting of the learning rate. So I think the lr should be zero even if in the first update. 

"Why don't you just set the values of important parameters to their original values after back-propagation without caring for their gradients?"

Yeah, this method is more convenient and can achieve the goal. Thank you!
